SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,215 @@
-			CompletedRequests.Add(EntityId);
-			continue;
-		}
+		ComponentIds.AddUnique(RequiredComponentId);
+	}
 
-		const SpatialGDK::NetOwningClientWorker* NetOwningClientWorkerComponent =
-			StaticComponentView->GetComponentData<SpatialGDK::NetOwningClientWorker>(EntityId);
-		if (NetOwningClientWorkerComponent == nullptr)
-		{
-			// This happens if the NetOwningClientWorker component is removed in the same tick as a request is queued, but the request was
-			// not removed from the queue - shouldn't happen.
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Error,
-				   TEXT("Cannot process entity as NetOwningClientWorker component has been removed since the request was queued. EntityId: "
-						"%lld"),
-				   EntityId);
-			CompletedRequests.Add(EntityId);
-			continue;
-		}
+	// Get the client worker ID net-owning this Actor from the NetOwningClientWorker.
+	const PhysicalWorkerName PossessingClientId =
+		NetOwningClientWorkerComponent.WorkerId.IsSet() ? NetOwningClientWorkerComponent.WorkerId.GetValue() : FString();
 
-		const SpatialGDK::ComponentPresence* ComponentPresenceComponent =
-			StaticComponentView->GetComponentData<SpatialGDK::ComponentPresence>(EntityId);
-		if (ComponentPresenceComponent == nullptr)
-		{
-			// This happens if the ComponentPresence component is removed in the same tick as a request is queued, but the request was not
-			// removed from the queue - shouldn't happen.
-			UE_LOG(
-				LogSpatialLoadBalanceEnforcer, Error,
-				TEXT("Cannot process entity as ComponentPresence component has been removed since the request was queued. EntityId: %lld"),
-				EntityId);
-			CompletedRequests.Add(EntityId);
-			continue;
-		}
+	const FString& WriteWorkerId = FString::Printf(TEXT("workerId:%s"), **DestinationWorkerId);
 
-		if (AuthorityIntentComponent->VirtualWorkerId == SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
-		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
-				   TEXT("Entity with invalid virtual worker ID assignment will not be processed. EntityId: %lld. This should not happen - "
-						"investigate if you see this warning."),
-				   EntityId);
-			CompletedRequests.Add(EntityId);
-			continue;
-		}
+	const WorkerAttributeSet OwningServerWorkerAttributeSet = { WriteWorkerId };
 
-		check(VirtualWorkerTranslator != nullptr);
-		const PhysicalWorkerName* DestinationWorkerId =
-			VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
-		if (DestinationWorkerId == nullptr)
+	for (const Worker_ComponentId& ComponentId : ComponentIds)
+	{
+		if (ComponentId == SpatialConstants::HEARTBEAT_COMPONENT_ID
+			|| ComponentId == SpatialConstants::GetClientAuthorityComponent(GetDefault<USpatialGDKSettings>()->UseRPCRingBuffer()))
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Error,
-				   TEXT("This worker is not assigned a virtual worker. This shouldn't happen! Worker: %s"), *WorkerId);
+			Acl.ComponentWriteAcl.Add(ComponentId, { { PossessingClientId } });
```

Comment:
Can we use else instead of continue?

Reply:
I just copied the ACL construction code over. It can probably be cleaned up, I'll TAL

Reply:
made it a switch! I like it a lot more now

Reply:
switch is bugged. Will fix